src/Sarif/Visitors/InsertOptionalDataVisitor.cs

```diff
@@ -161,18 +149,20 @@ public override Message VisitMessage(Message node)
                 _dataToInsert.Includes(OptionallyEmittedData.FlattenedMessages))
             {
                 Rule rule = null;
-                //string formatString = null;
-                if ((bool)_run.Resources?.Rules.TryGetValue(_ruleId, out rule))
+                string formatString = null;
+                if (_ruleId != null &&
+                    _run.Resources?.Rules.TryGetValue(_ruleId, out rule) == true)
+                {
+                    node.Text = node.Arguments?.Count > 0 
+                        ? rule.Format(node.MessageId, node.Arguments.ToArray()) 
```

Comment:
>Format [](start = 31, length = 6)

It occurs to me that, since we own the method `Rule.Format`, we could have it accept the arguments as `IEnumerable` rather than as an array. Then you wouldn't have to do the `ToArray` in each caller, you could do it once in `Rule.Format`

The downside is that all the .NET `Format` methods we know and love take a `params object[]`, so callers probably wouldn't think to use the labor-saving `IEnumerable` feature and would continue to call `ToArray`.

Up to you.

Reply:
we should fix it.

---
In reply to: [208325436](https://github.com/Microsoft/sarif-sdk/pull/992#discussion_r208325436) [](ancestors = 208325436)

Reply:
Change looks good.

---
In reply to: [208381881](https://github.com/Microsoft/sarif-sdk/pull/992#discussion_r208381881) [](ancestors = 208381881,208325436)